src/components/ThemeProvider.tsx

```diff
@@ -0,0 +1,132 @@
+import { ScriptOnce } from '@tanstack/react-router'
+import * as React from 'react'
+import { createContext, ReactNode, useEffect, useState } from 'react'
+import { z } from 'zod'
+
+const themeModeSchema = z.enum(['light', 'dark', 'auto'])
+const resolvedThemeSchema = z.enum(['light', 'dark'])
+const themeKey = 'theme'
+
+type ThemeMode = z.infer<typeof themeModeSchema>
+type ResolvedTheme = z.infer<typeof resolvedThemeSchema>
+
+function getStoredThemeMode(): ThemeMode {
+  if (typeof window === 'undefined') return 'auto'
+  try {
+    const storedTheme = localStorage.getItem(themeKey)
+    return themeModeSchema.parse(storedTheme)
+  } catch {
+    return 'auto'
+  }
+}
+
+function setStoredThemeMode(theme: ThemeMode): void {
+  if (typeof window === 'undefined') return
+  try {
+    const parsedTheme = themeModeSchema.parse(theme)
+    localStorage.setItem(themeKey, parsedTheme)
+  } catch {}
+}
+
+function getSystemTheme(): ResolvedTheme {
+  if (typeof window === 'undefined') return 'light'
+  return window.matchMedia('(prefers-color-scheme: dark)').matches
+    ? 'dark'
+    : 'light'
+}
+
+function updateThemeClass(themeMode: ThemeMode) {
+  const root = document.documentElement
+  root.classList.remove('light', 'dark', 'auto')
+  const newTheme = themeMode === 'auto' ? getSystemTheme() : themeMode
+  root.classList.add(newTheme)
+
+  if (themeMode === 'auto') {
+    root.classList.add('auto')
+  }
+}
+
+function setupPreferredListener() {
+  const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)')
+  const handler = () => updateThemeClass('auto')
+  mediaQuery.addEventListener('change', handler)
+  return () => mediaQuery.removeEventListener('change', handler)
+}
+
+const themeDetectorScript = (function () {
+  function themeFn() {
+    try {
+      const storedTheme = localStorage.getItem('theme') || 'auto'
+
+      if (storedTheme === 'auto') {
+        const autoTheme = window.matchMedia('(prefers-color-scheme: dark)')
```

Comment:
do we have a subscription to theme change on the client? e.g. if i switch from dark to light mode on the client in my browser or OS settings, does this update the theme here as well?